packages/cli-kit/src/session/exchange.ts

```diff
@@ -152,42 +159,47 @@ async function requestAppToken(
   if (api === 'admin' && store) {
     identifier = `${store}-${appId}`
   }
-  const appToken = await tokenRequest(params).then(buildApplicationToken)
+  const tokenResult = await tokenRequest(params)
+  const value = tokenResult.mapError(tokenRequestErrorHandler).valueOrThrow()
+  const appToken = await buildApplicationToken(value)
   return {[identifier]: appToken}
 }
 
-// eslint-disable-next-line @typescript-eslint/no-explicit-any
-async function tokenRequest(params: {[key: string]: string}, throws = true): Promise<any> {
+interface TokenRequestResult {
+  access_token: string
+  expires_in: number
+  refresh_token: string
+  scope: string
+}
+
+function tokenRequestErrorHandler(error: string) {
+  if (error === 'invalid_grant') {
+    // There's an scenario when Identity returns "invalid_grant" when trying to refresh the token
+    // using a valid refresh token. When that happens, we take the user through the authentication flow.
+    throw new InvalidGrantError()
+  }
+  if (error === 'invalid_request') {
+    // There's an scenario when Identity returns "invalid_request" when exchanging an identity token.
+    // This means the token is invalid. We clear the session and throw an error to let the caller know.
+    throw InvalidIdentityError
+  }
+  throw new Abort(error)
+}
+
+async function tokenRequest(params: {[key: string]: string}): Promise<Result<TokenRequestResult, string>> {
   const fqdn = await identityFqdn()
   const url = new URL(`https://${fqdn}/oauth/token`)
   url.search = new URLSearchParams(Object.entries(params)).toString()
   const res = await shopifyFetch(url.href, {method: 'POST'})
   // eslint-disable-next-line @typescript-eslint/no-explicit-any
   const payload: any = await res.json()
-  if (!res.ok) {
-    if (payload.error === 'invalid_grant') {
-      // There's an scenario when Identity returns "invalid_grant" when trying to refresh the token
-      // using a valid refresh token. When that happens, we take the user through the authentication flow.
-      throw new InvalidGrantError(payload.error_description)
-    } else if (payload.error === 'invalid_request') {
-      // There's an scenario when Identity returns "invalid_request" when exchanging an identity token.
-      // This means the token is invalid. We clear the session and throw an error to let the caller know.
-      await secureStore.remove()
-      throw InvalidIdentityError
-    } else {
-      if (!throws) return payload
-      throw new Abort(payload.error_description)
-    }
-  }
-  return payload
+
+  if (res.ok) return ok(payload)
+  if (payload.error === 'invalid_request') await secureStore.remove()
```

Comment:
The `await secureStore.remove()` side effect feels very deep in the logic. What about letting the logic that coordinates everything to deal with it?

Reply:
I thought of moving it to the errorHandler, but didn't want to make that function async since it's just a mapper... I'll take another look 👌